modules/preloader/injected/EthereumProvider.js

```diff
@@ -1,191 +1,201 @@
-      }, 500);
-    });
-
-    this.on('connect', function() {
-      clearInterval(reconnectIntervalId);
-    });
-  };
-
-  /**
-     Will try to make a connection
-
-     @method connect
-     */
-  EthereumProvider.prototype._connect = function(payload, callback) {
-    postMessage({
-      type: 'create'
-    });
-  };
-
-  /**
-     Sends the request
-
-     @method send
-     @param {Object} payload    example: {id: 1, jsonrpc: '2.0', 'method': 'eth_someMethod', params: []}
-     @param {Function} callback the callback to call
-     */
-  // TODO transform to: send(method, params, callback)
-  EthereumProvider.prototype.send = function send(payload, callback) {
-    this._addResponseCallback(payload, callback);
-    postMessage(
-      {
-        type: 'write',
-        message: payload
-      },
-      this.origin
-    );
-  };
-
-  delete window.EventEmitter;
-  // TODO set real ethereum provider
-  // window.ethereum = new EthereumProvider();
-
-  // For backwards compatibility of web3.currentProvider;
-  EthereumProvider.prototype.sendSync = function() {
-    return {
-      jsonrpc: '2.0',
-      error: {
-        code: -32603,
-        message: 'Sync calls are not anymore supported in Mist :\\'
-      }
-    };
-  };
-  EthereumProvider.prototype.sendAsync = EthereumProvider.prototype.send;
-  EthereumProvider.prototype.isConnected = function() {
-    return true;
-  };
-  window.web3 = {
-    currentProvider: new EthereumProvider()
-  };
+    /* Events */
+
+    _emitConnect() {
+      this._isConnected = true;
+      this.emit('connect');
+    }
+
+    _emitClose(code, reason) {
+      this._isConnected = false;
+      this.emit('close', code, reason);
+
+      // Send Error objects to any open subscriptions
+      this._activeSubscriptions.forEach(id => {
+        const error = new Error(
+          `Provider connection to network closed.
+           Subscription lost, please subscribe again.`
+        );
+        this.emit(id, error);
+      });
+      // Clear subscriptions
+      this._activeSubscriptions = [];
+    }
+
+    _emitNetworkChanged(networkId) {
+      this.emit('networkChanged', networkId);
+    }
+
+    _emitAccountsChanged(accounts) {
+      this.emit('accountsChanged', accounts);
+    }
+
+    /* web3.js provider compatibility */
+
+    sendAsync(payload, callback) {
+      return this.send(payload.method, payload.params)
+        .then(result => {
+          const response = payload;
+          response.result = result;
+          callback(null, response);
+        })
+        .catch(error => {
+          callback(error, null);
+          // eslint-disable-next-line no-console
+          console.error(
+            `Error from EthereumProvider sendAsync ${payload}: ${error}`
+          );
+        });
+    }
+
+    isConnected() {
+      return this._isConnected;
+    }
+
+    /* Deprecated methods */
+
+    sendSync() {
+      return new Error('Sync calls are not supported.');
+    }
+
+    sendBatch() {
```

Comment:
Batch calls are support but the sendBatch method not.. you could send a batch request payload over the normal send method. I think the error message should be something like: "sendBatch method is deprecated please use the send method for batch calls (URL to an info page)"